state-recover/appcore/clients-interfaces/src/main/kotlin/build/linea/staterecover/clients/LineaSubmissionEventsClient.kt

```diff
@@ -0,0 +1,143 @@
+    if (!finalStateRootHash.contentEquals(other.finalStateRootHash)) return false
+
+    return true
+  }
+
+  override fun hashCode(): Int {
+    var result = parentShnarf.contentHashCode()
+    result = 31 * result + shnarf.contentHashCode()
+    result = 31 * result + finalStateRootHash.contentHashCode()
+    return result
+  }
+
+  override fun toString(): String {
+    return "DataSubmittedV3(" +
+      "parentShnarf=${parentShnarf.encodeHex()}," +
+      " shnarf=${shnarf.encodeHex()}," +
+      " finalStateRootHash=${finalStateRootHash.encodeHex()})"
+  }
+}
+
+data class DataFinalizedV3(
+  override val startBlockNumber: ULong,
+  override val endBlockNumber: ULong,
+  val snarf: ByteArray,
+  val parentStateRootHash: ByteArray,
+  val finalStateRootHash: ByteArray
+) : BlockInterval {
+  companion object {
+    fun fromEthLog(ethLog: EthLog): EthLogEvent<DataFinalizedV3> {
+      /**event DataFinalizedV3(
+       uint256 indexed startBlockNumber,
+       uint256 indexed endBlockNumber,
+       bytes32 indexed shnarf,
+       bytes32 parentStateRootHash,
+       bytes32 finalStateRootHash
+       );*/
+      val dataBytes = ethLog.data
+      return EthLogEvent(
+        event = DataFinalizedV3(
+          startBlockNumber = ethLog.topics[1].toULongFromLast8Bytes(),
+          endBlockNumber = ethLog.topics[2].toULongFromLast8Bytes(),
+          snarf = ethLog.topics[3],
+          parentStateRootHash = dataBytes.sliceOf32(sliceNumber = 0),
+          finalStateRootHash = dataBytes.sliceOf32(sliceNumber = 1)
+        ),
+        log = ethLog
+      )
+    }
+  }
+
+  override fun equals(other: Any?): Boolean {
+    if (this === other) return true
+    if (javaClass != other?.javaClass) return false
+
+    other as DataFinalizedV3
+
+    if (startBlockNumber != other.startBlockNumber) return false
+    if (endBlockNumber != other.endBlockNumber) return false
+    if (!snarf.contentEquals(other.snarf)) return false
+    if (!parentStateRootHash.contentEquals(other.parentStateRootHash)) return false
+    if (!finalStateRootHash.contentEquals(other.finalStateRootHash)) return false
+
+    return true
+  }
+
+  override fun hashCode(): Int {
+    var result = startBlockNumber.hashCode()
+    result = 31 * result + endBlockNumber.hashCode()
+    result = 31 * result + snarf.contentHashCode()
+    result = 31 * result + parentStateRootHash.contentHashCode()
+    result = 31 * result + finalStateRootHash.contentHashCode()
+    return result
+  }
+
+  override fun toString(): String {
+    return "DataFinalizedV3(" +
+      "startBlockNumber=$startBlockNumber, " +
+      "endBlockNumber=$endBlockNumber, " +
+      "snarf=${snarf.encodeHex()}, " +
+      "parentStateRootHash=${parentStateRootHash.encodeHex()}, " +
+      "finalStateRootHash=${finalStateRootHash.encodeHex()})"
+  }
+}
+
+data class FinalizationAndDataEventsV3(
+  val dataSubmittedEvents: List<EthLogEvent<DataSubmittedV3>>,
+  val dataFinalizedEvent: EthLogEvent<DataFinalizedV3>
```

Comment:
> I assume dataSubmittedEvents is an ordered list and it includes all the data submitted events for the finalized event?

Yes, that is correct, atm at least.

> If that's the case, should we add required checks to make sure:

IMO no. That is the responsibility of the Linea protocol itself and the events fetcher class to fetch them properly. The data class is just a data holder. Putting those constraints is bringing the awareness/concern of order and relation between them to a 3rd place. 

Alos, mixing data holding and validations will require complex logic to create some data for testing because it needs to match those constraint's.